### PR TITLE
opt/sql: support ORDER BY INDEX, change behaviour

### PR DESCRIPTION
**Pull request by justinj**
This commit introduces support for ORDER BY INDEX into the optimizer. It
also slightly changes the behaviour of ORDER BY INDEX in the heuristic
planner.

Previously, ORDER BY INDEX would imply an ordering only by the columns
explicitly specified in the definition of the index. Now, we add the
implicitly added primary key columns as well.

It's not clear if the previous behaviour was more desirable (and I don't
have a strong opinion) but it has a handful of desirable properties:
* It doesn't require extending any of the interfaces in the opt package
  (namely, otherwise we would need to extend opt.Index with something like
  `ExplicitColumns`),
* it now meets the stated goal of guaranteeing an ordering (because
  every such ordering now contains a key),
* it's a strengthening of the previous behaviour, the semantics
  guaranteed before are still guaranteed.

However, I can see some downsides:
* It requires fetching more columns in some cases,
* it is a stronger guarantee,
* it leaks the way we order indexes internally.

Release note (sql change): ORDER BY INDEX now implies an ordering by the
implicit primary key columns appended to indexes.

### CODE REVIEW

**Review by andy-kimball — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/optbuilder/orderby.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LHsvuR-21Q95Z2bNhjK:-LHsvuR-21Q95Z2bNhjL:b-5vae4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c884eda8fdd30a7abf27076a1a688a486e938ce6/pkg/sql/opt/optbuilder/orderby.go#L114)):*
> ```Go
> 		// Force the name to be fully qualified so that the column references
> 		// below only match against real tables in the FROM clause.
> 		tn.ExplicitCatalog = true
> ```

I'd thought that `resolveTable` was supposed to update the `tableName` with the fully-qualified name? It seems strange to be doing it manually here.
___
*[pkg/sql/opt/optbuilder/testdata/orderby, line 989 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LHsu7f09jNkYws28CP8:-LHsu7f09jNkYws28CP9:by8a04r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c884eda8fdd30a7abf27076a1a688a486e938ce6/pkg/sql/opt/optbuilder/testdata/orderby#L989)):*
> ```
> SELECT a FROM abcd ORDER BY INDEX abcd@nonexistent
> ----
> error: index "nonexistent" not found
> ```

I'd add a test case where input is not a table, but instead a sub-select like `(SELECT b FROM abcd) AS abcd`. I'd be trying to construct cases where all the index columns are not projected by the sub-select.

Also, you need to test a few qualified cases, like `public.abcde` and `mydb.public.abcde`.

**Review by knz — APPROVED** [review]
LGTM modulo resolution of Andy's concerns. Also a nit.

> Reviewed 6 of 7 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-:-LI6NiqU9xpAlY91-u4w:bog204u)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sort.go, line 444 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI6NCEd1vJp7X5FtWu_:-LI6NCEe1F0jkMq09xTX:b-uabj7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0982d94ecb92037c57caef2c1cd4285180c5bad/pkg/sql/sort.go#L444)):*
> ```Go
> 				col, err := desc.FindColumnByID(id)
> 				if err != nil {
> 					return nil, errors.Errorf("programming error: column with ID %d not found", id)
> ```

`pgerror.NewErrorf(pgerror.CodeInternalError, ...)`

**Review by justinj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sort.go, line 444 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI6NCEd1vJp7X5FtWu_:-LI6WtSf8c4Bss49omQp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0982d94ecb92037c57caef2c1cd4285180c5bad/pkg/sql/sort.go#L444)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`pgerror.NewErrorf(pgerror.CodeInternalError, ...)`
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/orderby.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LHsvuR-21Q95Z2bNhjK:-LI6Wi3M7EwvZaqBwu8J:bcxumea) ([raw file](https://github.com/cockroachdb/cockroach/blob/c884eda8fdd30a7abf27076a1a688a486e938ce6/pkg/sql/opt/optbuilder/orderby.go#L114)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd thought that `resolveTable` was supposed to update the `tableName` with the fully-qualified name? It seems strange to be doing it manually here.
</blockquote></details>

Ah, I think that was the intention (the comment on `TabName` implies as such) but that wasn't the behaviour. I've moved this into the right place now.
___
*[pkg/sql/opt/optbuilder/testdata/orderby, line 989 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LHsu7f09jNkYws28CP8:-LI6XC3t8Ew3J9eZLMa2:b1vu6ao) ([raw file](https://github.com/cockroachdb/cockroach/blob/c884eda8fdd30a7abf27076a1a688a486e938ce6/pkg/sql/opt/optbuilder/testdata/orderby#L989)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd add a test case where input is not a table, but instead a sub-select like `(SELECT b FROM abcd) AS abcd`. I'd be trying to construct cases where all the index columns are not projected by the sub-select.

Also, you need to test a few qualified cases, like `public.abcde` and `mydb.public.abcde`.
</blockquote></details>

I added one of those, but those are just not valid - the table must be a direct data source (since the generated column names become fully qualified).

The grammar only allows specifying a single unqualified name after the `@`, but I added a couple qualified table names.

**Review by RaduBerinde — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-:-LI6d33fAyefeJ5oCGEC:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by rytaft — CHANGES_REQUESTED** [review]
:lgtm: 

[nit] This sentence in the commit message is a bit ambiguous:
> It's not clear if the previous behaviour was more desirable (and I don't
> have a strong opinion) but it has a handful of desirable properties:

I'd change it to:
> It's not clear if the previous behaviour was more desirable (and I don't
> have a strong opinion) but the new behavior has a handful of desirable
> properties:

> Reviewed 4 of 7 files at r1, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-:-LI7gRPA6HExNkEqrblA:b-8nxzjp)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/logictest/testdata/logic_test/order_by, line 444 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7h3twAfGFb2CrcW8Q:-LI7h3twAfGFb2CrcW8R:br9op77) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/logictest/testdata/logic_test/order_by#L444)):*
> ```
> 1  30  500
> 
> statement error does not exist
> ```

[nit] this error message seems clipped - would help to include the context of the non-existent name (ditto the next one)
___
*[pkg/sql/opt/optbuilder/orderby.go, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7jaUi9Nfufz7U0rCT:-LI7jaUi9Nfufz7U0rCU:buol0ar) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/opt/optbuilder/orderby.go#L76)):*
> ```Go
> 
> // addOrderingColumn adds expr as an ordering column to orderByScope, if it is
> // already rendered in projectionsScope then that render is re-used.
> ```

[nit] we aren't really using the term "render" in the optimizer -- I'd rephrase this to say "projection"
___
*[pkg/sql/opt/optbuilder/orderby.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7kV5sCzz1TyDjwT38:-LI7kV5sCzz1TyDjwT39:bj1sse5) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/opt/optbuilder/orderby.go#L116)):*
> ```Go
> 		}
> 
> 		expr := inScope.resolveType(tree.NewColumnItem(tab.TabName(), tree.Name(col.Column.ColName())), types.Any)
> ```

Why do we need to `resolveType` here? Don't we already know everything we need to know from `resolveTable` and `findIndexByName`? I think you can just create a `scopeColumn` directly using that info... (take a look at `buildScan` in `optbuilder/select.go` for an example.)

**Review by justinj — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/sql/logictest/testdata/logic_test/order_by, line 444 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7h3twAfGFb2CrcW8Q:-LIHU5ys8-RxS4QW6yI6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/logictest/testdata/logic_test/order_by#L444)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] this error message seems clipped - would help to include the context of the non-existent name (ditto the next one)
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/orderby.go, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7jaUi9Nfufz7U0rCT:-LIHU6xX1s1mKiRuH-wd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/opt/optbuilder/orderby.go#L76)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] we aren't really using the term "render" in the optimizer -- I'd rephrase this to say "projection"
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/orderby.go, line 116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7kV5sCzz1TyDjwT38:-LII-jwNC5LVMzb4bmiN:b-n8unqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/opt/optbuilder/orderby.go#L116)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why do we need to `resolveType` here? Don't we already know everything we need to know from `resolveTable` and `findIndexByName`? I think you can just create a `scopeColumn` directly using that info... (take a look at `buildScan` in `optbuilder/select.go` for an example.)
</blockquote></details>

We discussed this offline - going to leave it as it is.

**Review by rytaft — APPROVED** [review]
> Reviewed 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27812)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/sql/logictest/testdata/logic_test/order_by, line 444 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/27812#-LI7h3twAfGFb2CrcW8Q:-LII8GuB0_pAvXGgoxYy:b4axv65) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d718ba7f472c79d168f55be867238853b01c878/pkg/sql/logictest/testdata/logic_test/order_by#L444)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can you also update the next one too? (no data source matches)

**Comment by justinj** [comment]
Fixed! TFTRs!

bors r+
